apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/shard/helper/SecondaryIndexSelectAndInsertHelper.java

```diff
@@ -0,0 +1,165 @@
+            do {
+                if ("BLOCK_INDEX".equalsIgnoreCase(currentTableName)) {
+                    ps.setLong(1, operationParams.shardId.get());
+                    ps.setLong(2, paginateResultWrapper.limitValue);
+                    ps.setLong(3, upperBoundIdValue);
+                    ps.setLong(4, operationParams.batchCommitSize);
+                } else if ("TRANSACTION_SHARD_INDEX".equalsIgnoreCase(operationParams.tableName)) {
+                    ps.setLong(1, paginateResultWrapper.limitValue);
+                    ps.setLong(2, upperBoundIdValue);
+                    ps.setLong(3, operationParams.batchCommitSize);
+                }
+            } while (handleResultSet(ps, paginateResultWrapper, sourceConnect, operationParams.batchCommitSize));
+        } catch (Exception e) {
+            log.error("Processing failed, Table " + currentTableName, e);
+            throw e;
+        } finally {
+            if (this.preparedInsertStatement != null && !this.preparedInsertStatement.isClosed()) {
+                this.preparedInsertStatement.close();
+            }
+        }
+        log.debug("Inserted '{}' = [{}] within {} secs", operationParams.tableName, totalRowCount, (System.currentTimeMillis() - startSelect) / 1000);
+
+        // turn ON HEIGHT constraint for specified table
+        if ("BLOCK_INDEX".equalsIgnoreCase(currentTableName)) {
+            issueConstraintUpdateQuery(sourceConnect,
+                    "CREATE UNIQUE INDEX IF NOT EXISTS block_index_block_id_shard_id_idx ON block_index (block_id, shard_id DESC)");
+            issueConstraintUpdateQuery(sourceConnect,
+                    "CREATE UNIQUE INDEX IF NOT EXISTS block_index_block_height_shard_id_idx ON block_index (block_height, shard_id DESC)");
+        } else if ("TRANSACTION_SHARD_INDEX".equalsIgnoreCase(currentTableName)) {
+            issueConstraintUpdateQuery(sourceConnect,
+                    "ALTER TABLE transaction_shard_index ADD CONSTRAINT IF NOT EXISTS transaction_shard_index_block_fk FOREIGN KEY (block_id) REFERENCES block_index(block_id)");
+            issueConstraintUpdateQuery(sourceConnect,
+                    "CREATE UNIQUE INDEX IF NOT EXISTS transaction_index_shard_1_idx ON transaction_shard_index (transaction_id, block_id)");
+        }
+
+        log.debug("Total (with CONSTRAINTS) '{}' = [{}] in {} secs", operationParams.tableName, totalRowCount, (System.currentTimeMillis() - startSelect) / 1000);
+        return totalRowCount;
+    }
+
+    private boolean handleResultSet(PreparedStatement ps, PaginateResultWrapper paginateResultWrapper,
+                                    Connection targetConnect, long batchCommitSize)
+            throws SQLException {
+        int rows = 0;
+        try (ResultSet rs = ps.executeQuery()) {
+            while (rs.next()) {
+                /*
+                 * handle rows here
+                 */
+                if (rsmd == null) {
```

Comment:
Can we move this code to another method outside the main loop? Also seems like this code is repeatable.

Reply:
moved to private, that code works one time only